[lldb][NativePDB] Fix crash in debugger when PDB has bad type index value

gv opened the pull request:
Fix crash when an inline site record in the PDB file contains type index which is out of bounds

gv committed:
[lldb][NativePDB] Fix crash in debugger when PDB has bad type index value

Fix crash when an inline site record in the PDB file contains type
index which is not an index of a record in the IPI stream

JDevlieghere commented:
Can you add a test that (used to) reproduce(s) the crash? Presumably this isn't covered when running the API tests with PDB support?

Nerixyz commented on llvm/lib/DebugInfo/CodeView/LazyRandomTypeCollection.cpp:
```diff
@@ -93,20 +93,28 @@ CVType LazyRandomTypeCollection::getType(TypeIndex Index) {
+  llvm::Expected<CVType> res = getTypeOrError(Index);
+  if (!res)
+    return std::nullopt;
+  return *res;
```
You can use `llvm::expectedToOptional` here.

Nerixyz reviewed (commented):
The LLDB changes look good to me. Though, it looks like the LLVM tests are failing. It might be easier if you land the LLVM changes first.

> Can you add a test that (used to) reproduce(s) the crash? Presumably this isn't covered when running the API tests with PDB support?

This feels similar to #166090 (i.e. an artifact of incremental linking), so testing is a bit more complex. yaml2pdb is probably the way to go here. Currently, that won't work, because the generated PDB is missing the section headers. I'll try to add this, so both PRs can have a test.

Opened #166566 for this.

Nerixyz commented:
> The resulting file can be loaded into the debugger all right (which wasn't the case before) but it can't resolve any address, because it doesn't have any section contributions (that can be confirmed with `llvm-pdbutil dump --section-contribs`).

I'm surprised the debugger can even load the PDB. In my testing, section contributions weren't necessary.

I forgot to mention that this needs another patch - https://github.com/llvm/llvm-project/pull/168211.
The reason LLDB is crashing right now is that the offsets included in symbols from the modules (mainly `S_G/LPROC32`) are incorrect. And that's because the reconstructed symbols from YAML don't include all information. Specifically, `S_INLINESITE` has variable length annotation data, which gets lost and causes it to change size.
`S_G/LPROC32` contains an offset to where the function will end (other records include similar offsets iirc). Now that `S_INLINESITE` has changed size, this is incorrect. 

Arguably, we should also handle this gracefully.

gv committed:
Add regression test

gv commented:
> Can you add a test that (used to) reproduce(s) the crash? Presumably this isn't covered when running the API tests with PDB support?

Added a test that causes a crash in unpatched code. Done so per @Nerixyz's advice (i. e. took the YAML file from #166090 and changed some numbers). @JDevlieghere What do you think? Should I reorder the commits or it doesn't matter?

Nerixyz reviewed (approved):
The LLDB changes look good to me. I don't know if it's better to land the LLVM changes separately.

gv committed:
Fix formatting

gv committed:
Fix PdbAstBuilderClang::GetOrCreateClangDeclContextForUid for cases
where GetOrCreateDeclForUid returns empty CompilerDecl
